graylog2-server/src/main/java/org/graylog/scheduler/JobSchedulerService.java

```diff
@@ -75,9 +109,66 @@ protected void run() throws Exception {
         }
     }
 
+    @Subscribe
+    public void handleJobCompleted(JobCompletedEvent triggerCompletedEvent) {
+        // The job execution engine has just completed a job so we want to check for runnable triggers immediately.
+        sleeper.interrupt();
+    }
+
     @Override
     protected void triggerShutdown() {
+        // We don't want to process events when shutting down, so do this first
+        schedulerEventBus.unregister(this);
         shutdownLatch.countDown();
         jobExecutionEngine.shutdown();
     }
+
+    /**
+     * This class provides a sleep method that can be interrupted without interrupting threads.
+     * The same could be achieved by using a {@link CountDownLatch} but that one cannot be reused and we would need
+     * to create new latch objects all the time. This implementation is using a {@link Semaphore} internally which
+     * can be reused.
+     */
+    @VisibleForTesting
+    static class InterruptibleSleeper {
+
+        private final Semaphore semaphore;
+
+        InterruptibleSleeper() {
+            this(new Semaphore(1));
+        }
+
+        @VisibleForTesting
+        InterruptibleSleeper(Semaphore semaphore) {
+            // Semaphores with more than one permit would break the assumptions of the implementation
+            checkArgument(semaphore.availablePermits() == 1, "Semaphore must only have 1 permit");
+            this.semaphore = semaphore;
+        }
+
+        /**
+         * Blocks for the given duration or until interrupted via {@link #interrupt()}.
+         *
+         * @param duration the duration to sleep
+         * @param unit     the duration unit
+         * @return true if slept for the given duration, false if interrupted
+         * @throws InterruptedException if the thread gets interrupted
+         */
+        public boolean sleep(long duration, TimeUnit unit) throws InterruptedException {
+            // First try to acquire the one permit we allow
+            semaphore.tryAcquire();
+            // Now try to acquire another permit. This won't work except #interrupt() got called in the meantime.
+            // It waits for the given duration, basically emulating a sleep.
+            final boolean acquired = semaphore.tryAcquire(duration, unit);
+            // Always release so we can start with a clean slate when sleep is called again
+            semaphore.release();
+            return !acquired;
+        }
+
+        /**
+         * Interrupt a {@link #sleep(long, TimeUnit)} call so it unblocks.
+         */
+        public void interrupt() {
+            semaphore.release();
```

Comment:
This looks like it could "overrelease" permits because the `tryAcquire` and `release` calls are unbalanced: there are more jobs that complete than sleeps.

Reply:
Good catch, thank you! :+1: We fixed it by draining all permits before the `tryAcquire(timeout)`.